### PR TITLE
Alter definite assignment for async and iterator local functions

### LINKED ISSUE #13762: Roslyn fails to enforce definite assignment rule for a variable assigned in a local iterator function or local async function
Roslyn fails to enforce definite assignment rules for variables which are assigned in local iterator functions or local async functions which are called before the variables in question are used.

Since these types of local functions don't execute all statements at the first call it can't be proven that assignment was executed.

here are some examples:

1) Since I don't iterate the result of Local function it's body won't execute at all and I will end up using unassigned local variable i;
```cs
using System;
using System.Collections.Generic;

public class C
{
  static void Main()
  {
    int i;

    Local();

    Console.WriteLine(i);

    IEnumerable<int> Local()
    {
      yield return 0;

      i = 10;
    }
  }
}
```

2) This case even marks statement 'i = 10;' as unreachable but still allows to use the variable in the enclosing method.
```cs
using System;
using System.Collections.Generic;

public class C
{
  static void Main()
  {
    int i;

    Local();

    Console.WriteLine(i);

    IEnumerable<int> Local()
    {
      yield break;

      i = 10;
    }
  }
}
```

3) It's almost guaranteed that local variable i won't be assigned by the time it's used.
```cs
using System;
using System.Collections.Generic;
using System.Threading.Tasks;

public class C
{
  static void Main()
  {
    int i;

    Local();

    Console.WriteLine(i);

    async Task Local()
    {
      await Task.Delay(1000*60*60);
      i = 10;
    }
  }
}
```

### PR DESCRIPTION
**Pull request by agocke**
This PR both adds the missing branches (yield/await) for async and iterator
local functions and disables allowing async and iterator local functions
definitely assigning captured variables.

Fixes #13762

### CODE REVIEW

File: src/Compilers/CSharp/Test/Semantic/FlowAnalysis/LocalFunctions.cs [hunk 1]
```diff
@@ -7,6 +8,136 @@ namespace Microsoft.CodeAnalysis.CSharp.UnitTests
...
+            var comp = CreateCompilationWithMscorlib46(@"
+using System.Collections.Generic;
+
+class C
+{
+    public static void M2()
+    {
+        int w, x, y, z;
+
+        L1();
```
Comment (jaredpar) [inline src/Compilers/CSharp/Test/Semantic/FlowAnalysis/LocalFunctions.cs]: Consider adding a test to verify that an iterator / async doesn't unset the definite assignment state of variables it writes.  

```csharp
int x;
x = 0;
L1();
Use(x);  
IEnumerable<int> L1() {
  yield break;
}
```

**Comment by gafter** [comment]
I don't think this is the right way to handle an async function. For replaying writes, the set of writes caused by a call to an async function should be the set of variables that are definitely assigned at every await expression, return statement, and endpoint of the async function.

---
In reply to: [249773991](https://github.com/dotnet/roslyn/pull/14096#issuecomment-249773991) [](ancestors = 249773991)

File: src/Compilers/CSharp/Test/Semantic/FlowAnalysis/LocalFunctions.cs [hunk 2]
```diff
@@ -7,6 +8,136 @@ namespace Microsoft.CodeAnalysis.CSharp.UnitTests
...
+            x = 0;
+            await Task.Delay(0);
+            y = 0;
+        }
+    }
+}");
+            comp.VerifyDiagnostics(
+                // (10,9): error CS0165: Use of unassigned local variable 'x'
+                //         x++;
+                Diagnostic(ErrorCode.ERR_UseDefViolation, "x").WithArguments("x").WithLocation(10, 9),
```
Comment (gafter) [inline src/Compilers/CSharp/Test/Semantic/FlowAnalysis/LocalFunctions.cs]: >ERR_UseDefViolation [](start = 37, length = 19)

I think this error is incorrect. Control does not return to the caller until `x` has been assigned.

**Comment by agocke** [comment]
@gafter This is the change I mentioned in person last week -- I'm planning to make the more conservative change and I'll send out an email to the LDM giving a brief synopsis of the situation. If we decide to relax the requirements, I just need to delete one line. :)

**Comment by agocke** [comment]
@gafter Updated. Should match your and the LDM's guidance now :)

File: src/Compilers/CSharp/Portable/FlowAnalysis/DataFlowPass.LocalFunctions.cs [hunk 3]
```diff
@@ -31,6 +32,7 @@ private class LocalFuncUsages
         {
             _usedLocalFunctions.Add(localFunc);
 
+
```
Comment (gafter) [inline src/Compilers/CSharp/Portable/FlowAnalysis/DataFlowPass.LocalFunctions.cs]: I think you unintentionally added an empty line here.

**Comment by agocke** [comment]
test windows_debug_unit32_prtest please

**Comment by agocke** [comment]
retest windows_debug_unit32_prtest please

**Pull request merged**